integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/testsupport/util/extension/SearchSetupHelper.java

```diff
@@ -37,42 +37,101 @@
 import org.hibernate.search.util.common.logging.impl.LoggerFactory;
 import org.hibernate.search.util.impl.integrationtest.common.TestConfigurationProvider;
 import org.hibernate.search.util.impl.integrationtest.common.bean.ForbiddenBeanProvider;
+import org.hibernate.search.util.impl.integrationtest.common.extension.ComposedExtension;
 import org.hibernate.search.util.impl.integrationtest.mapper.stub.StubMappedIndex;
 import org.hibernate.search.util.impl.integrationtest.mapper.stub.StubMapping;
 import org.hibernate.search.util.impl.integrationtest.mapper.stub.StubMappingImpl;
 import org.hibernate.search.util.impl.integrationtest.mapper.stub.StubMappingInitiator;
 import org.hibernate.search.util.impl.integrationtest.mapper.stub.StubMappingKey;
 import org.hibernate.search.util.impl.integrationtest.mapper.stub.StubMappingSchemaManagementStrategy;
 
-import org.junit.rules.RuleChain;
-import org.junit.rules.TestRule;
-import org.junit.runner.Description;
-import org.junit.runners.model.Statement;
-
-public class SearchSetupHelper implements TestRule {
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.BeforeTestExecutionCallback;
+import org.junit.jupiter.api.extension.Extension;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.TestExecutionExceptionHandler;
+
+public class SearchSetupHelper
+		implements AfterAllCallback, AfterEachCallback, AfterTestExecutionCallback,
+		BeforeAllCallback, BeforeEachCallback, BeforeTestExecutionCallback, TestExecutionExceptionHandler {
+
+	private enum Type {
+		CLASS,
+		METHOD;
+	}
```

Comment:
hmm, I'm not sure that `@TestInstance` can work here... this `Type` ^ is to decide which callback pair will be called `*Each` or `*All` ... 
Or do you mean that we should read the lifecycle from the extension context and then use it instead of the custom enum? (not sure if that'll be simpler though... )

Reply:
> hmm, I'm not sure that `@TestInstance` can work here... this `Type` ^ is to decide which callback pair will be called `*Each` or `*All` ... 

Ah, right, this doesn't look related.

> Or do you mean that we should read the lifecycle from the extension context and then use it instead of the custom enum? (not sure if that'll be simpler though... )

> (not sure if that'll be simpler though... )

In the helpers maybe not. If it's simpler in tests themselves (i.e. no explicit `.create()`/`createGlobal()`, I'd still consider this a win though.

Perhaps you could just detect whether the `start` method gets called from a `@BeforeAll` method or from a `@BeforeEach`/`@Test` method?

* If it's called from a `@BeforeAll` method, you know any mapping created in that method will have to be cleaned up in `@AfterAll`.
* If it's called from a `@BeforeEach` or `@Test` method, you know any mapping created in that method will have to be cleaned up in `@AfterEach`.

Assuming the extension's own `beforeEachCallback` gets called before any `@BeforeEach`/`@Test` method, this could easily be achieved by updating some state in the helper when `beforeEachCallback` gets called.

Reply:
yeah, I was waiting for the parameterized class to be more or less finished before looking into this. I think that the "autodetection" that I've added in the new commit covers our needs and there's less manual work required 😃

Reply:
Yep, what you did for `SearchSetupHelper` looks good, thanks :+1: